core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1102,6 +1121,22 @@ public Builder setDefaultPartitionSpec(int specId) {
       return this;
     }
 
+    private Builder removePartitionSpec(PartitionSpec spec) {
+      Preconditions.checkArgument(
+          changes.isEmpty(), "Cannot remove partition spec with other metadata update");
+      Preconditions.checkArgument(
+          spec.specId() != defaultSpecId, "Cannot remove default partition spec");
+      PartitionSpec toBeRemoved = specsById.remove(spec.specId());
+      Preconditions.checkArgument(
+          toBeRemoved != null && toBeRemoved.equals(spec),
+          "Cannot remove an unknown spec, spec id: %s",
+          spec.specId());
+      this.specs =
```

Comment:
Should we be validating that all the specs we are trying to remove exist? I think this may be fine, just thinking about commit conflicts

Reply:
In general I'm in favor of succeeding in cases like this. There's no need to fail something if it has an accidental retry after the first attempt succeeded.

Reply:
In that case, is it really necessary to validate that a spec was removed and filter the list? If it is a noop to remove an unknown spec ID, then we can simplify the first half of this method:

```java
    Preconditions.checkArgument(!specIds.contains(defaultSpecId), "Cannot remove the default partition spec");
```